src/Mafia/Bin.hs

```diff
@@ -0,0 +1,89 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PatternSynonyms #-}
+module Mafia.Bin
+  ( InstallPackage(..)
+  , ipackageId
+  , renderInstallPackage
+  , ipkgName
+  , ipkgVersion
+
+  , installBinary
+  , ensureExeOnPath
+  ) where
+
+import           Control.Monad.IO.Class (MonadIO(..))
+
+import           Data.Text (Text)
+import qualified Data.Text as T
+
+import           Mafia.Home
+import           Mafia.IO
+import           Mafia.Install
+import           Mafia.Package
+import           Mafia.Path
+import           P
+
+import           System.IO (IO)
+import           System.Posix.Files (createSymbolicLink)
+
+import           X.Control.Monad.Trans.Either (EitherT)
+
+
+data InstallPackage =
+    InstallPackageName PackageName
+  | InstallPackageId PackageId
+    deriving (Eq, Ord, Show)
+
+ipackageId :: Text -> [Int] -> InstallPackage
+ipackageId name ver =
+  InstallPackageId (packageId name ver)
+
+renderInstallPackage :: InstallPackage -> Text
+renderInstallPackage = \case
+  InstallPackageName name ->
+    unPackageName name
+  InstallPackageId pid ->
+    renderPackageId pid
+
+ipkgName :: InstallPackage -> PackageName
+ipkgName = \case
+  InstallPackageName name ->
+    name
+  InstallPackageId pid ->
+    pkgName pid
+
+ipkgVersion :: InstallPackage -> Maybe Version
+ipkgVersion = \case
+  InstallPackageName _ ->
+    Nothing
+  InstallPackageId pid ->
+    Just (pkgVersion pid)
+
+-- | Installs a given cabal package at a specific version and return a directory containing all executables
+installBinary :: InstallPackage -> EitherT InstallError IO Directory
+installBinary ipkg = do
+  bin <- ensureMafiaDir "bin"
+
+  let
+    plink = bin </> renderInstallPackage ipkg
+    pdir = plink <> "/"
+    pbin = plink <> "/bin"
+
+  unlessM (doesDirectoryExist pdir) $ do
+    -- if the directory doesn't exist, but there happens to be a file there, we
+    -- must have a dead symlink, so lets remove it and install it again.
+    ignoreIO $ removeFile plink
+
+    pkg <- installPackage (ipkgName ipkg) (ipkgVersion ipkg)
+    env <- getPackageEnv
+    let gdir = packageSandboxDir env pkg
+    liftIO $ createSymbolicLink (T.unpack gdir) (T.unpack plink)
+
+  return pbin
+
+ensureExeOnPath :: InstallPackage -> EitherT InstallError IO ()
+ensureExeOnPath pkg = do
+  dir <- installBinary pkg
+  setEnv "PATH" . maybe dir (\path -> dir <> ":" <> path) =<< lookupEnv "PATH"
```

Comment:
I know I added this, but I wish I'd done something a little safer - like creating a `Exe` data type which you pass around.